tests/models/layers/test_flash_attn.py

```diff
@@ -253,3 +254,107 @@ def test_sliding_window(sliding_window_size: int):
+                                                       device=device)
+    key_1.requires_grad = True
+    value_1 = torch.randn(bsz, seqlen_1, n_heads * d).to(dtype=dtype,
+                                                         device=device)
+    value_1.requires_grad = True
+    attn_bias_1 = gen_slopes(n_heads=n_heads,
+                             alibi_bias_max=8,
+                             device=torch.device(device),
+                             return_1d=True)
+    output_1, _, _ = flash_attn_fn(query=query_1,
+                                   key=key_1,
+                                   value=value_1,
+                                   n_heads=n_heads,
+                                   kv_n_heads=n_heads,
+                                   past_key_value=None,
+                                   softmax_scale=1 / math.sqrt(d),
+                                   attn_bias=attn_bias_1,
+                                   key_padding_mask=None,
+                                   is_causal=True,
+                                   dropout_p=0.0,
+                                   training=False,
+                                   needs_weights=False,
+                                   multiquery=False,
+                                   attention_mask_in_length=None,
+                                   should_repeat_kv_for_gqa=True)
+
+    output_1.sum().backward()
+
+    query_2 = query_1.detach().clone()
+    query_2.requires_grad = True
+    key_2 = key_1.detach().clone()
+    key_2.requires_grad = True
+    value_2 = value_1.detach().clone()
+    value_2.requires_grad = True
+
+    def gen_bias():
+        causal = True
+        bs = attention.attn_bias_shape('triton',
+                                       n_heads,
+                                       seqlen_1,
+                                       True,
+                                       prefix_lm=False,
+                                       use_sequence_id=False,
+                                       causal=causal)
+
+        attn_bias = torch.zeros(*bs, device=device)
+        attn_bias = attention.build_attn_bias(
+            'triton',
+            attn_bias,
+            n_heads,
+            seqlen_1,
+            causal=causal,
+            alibi=True,
+            alibi_bias_max=8,
+        )
+        return attn_bias
+
+    attn_bias_2 = gen_bias()
+
+    output_2, _, _ = triton_flash_attn_fn(
+        query=query_2,
+        key=key_2,
+        value=value_2,
+        n_heads=n_heads,
+        kv_n_heads=n_heads,
+        past_key_value=None,
+        softmax_scale=1 / math.sqrt(d),
+        attn_bias=attn_bias_2,
+        key_padding_mask=None,
+        is_causal=True,
+        dropout_p=0.0,
+        training=False,
+        needs_weights=False,
+        multiquery=False,
+    )
+
+    output_2.sum().backward()
+
+    assert torch.allclose(output_1, output_2)
+    assert torch.norm(query_2.grad - query_1.grad  # type: ignore
```

Comment:
what's the reason for the type ignore?

Reply:
pyright complained that `-` cannot operate between `None` type objects, because `query_1.grad` and` query_2.grad` can potentially be `None`.

Reply:
Fixed this.